bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -0,0 +1,358 @@
+#[derive(Debug)]
+struct AllowedRequests {
+    migrate_as_target: RequestDisposition,
+    start: RequestDisposition,
+    migrate_as_source: RequestDisposition,
+    reboot: RequestDisposition,
+    stop: RequestDisposition,
+}
+
+#[derive(Debug)]
+pub struct ExternalRequestQueue {
+    queue: VecDeque<ExternalRequest>,
+    allowed: AllowedRequests,
+    log: Logger,
+}
+
+impl ExternalRequestQueue {
+    /// Creates a new queue that logs to the supplied logger.
+    pub fn new(log: Logger) -> Self {
+        Self {
+            queue: VecDeque::new(),
+            allowed: AllowedRequests {
+                migrate_as_target: RequestDisposition::Enqueue,
+                start: RequestDisposition::Enqueue,
+                migrate_as_source: RequestDisposition::Deny(
+                    RequestDeniedReason::InstanceNotActive,
+                ),
+                reboot: RequestDisposition::Deny(
+                    RequestDeniedReason::InstanceNotActive,
+                ),
+                stop: RequestDisposition::Enqueue,
+            },
+            log,
+        }
+    }
+
+    /// Pops the request at the front of the queue.
+    pub fn pop_front(&mut self) -> Option<ExternalRequest> {
+        self.queue.pop_front()
+    }
+
+    /// Indicates whether the queue is empty.
+    pub fn is_empty(&self) -> bool {
+        self.queue.is_empty()
+    }
+
+    /// Asks to place the supplied request on the queue.
+    pub fn try_queue(
+        &mut self,
+        request: ExternalRequest,
+    ) -> Result<(), RequestDeniedReason> {
+        use RequestDeniedReason as DenyReason;
+        use RequestDisposition as Disposition;
+
+        let disposition = match request {
+            ExternalRequest::MigrateAsTarget { .. } => {
+                self.allowed.migrate_as_target
+            }
+            ExternalRequest::Start => self.allowed.start,
+            ExternalRequest::MigrateAsSource { .. } => {
+                self.allowed.migrate_as_source
+            }
+            ExternalRequest::Reboot => self.allowed.reboot,
+
+            // Requests to stop always succeed. Note that a request to stop a VM
+            // that hasn't started should still be queued to the state worker so
+            // that the worker can exit and drop its references to the instance.
+            ExternalRequest::Stop => self.allowed.stop,
+        };
+
+        match disposition {
+            Disposition::Enqueue => {}
+            Disposition::Ignore => return Ok(()),
+            Disposition::Deny(reason) => return Err(reason),
+        };
+
+        info!(&self.log, "queuing external request";
+              "request" => ?request,
+              "disposition" => ?disposition);
+
+        // At this point the request will be queued. Queuing some requests
+        // logically forecloses on other kinds of requests. Update the
+        // dispositions of these requests, then queue the request.
+        match request {
```

Comment:
When reading this piece of this method, I found myself wanting a table helping me translate the request and the set of allowed states after we enqueue it. This was particularly true where the resulting disposition of a request type would be implicit.

What would you think about matching on the request and returning from the `match` the AllowedRequests struct that would be associated with it? That might make it easier to grok reading this, since you can walk up to the function and quickly glean what the resulting dispositions will be.

For example, for the `MigrateAsTarget` case, something like:
```
let allowed_requests = match request {
    ExternalRequest::MigrateAsTarget { .. } => {
        AllowedRequests {
            migrate_as_target: Disposition::Ignore,
            start: Disposition::Ignore,
            migrate_as_source: Disposition::Deny(DenyReason::MigrateTargetInProgress),
            reboot: Disposition::Deny(DenyReason::MigrateTargetInProgress),
            stop: Disposition::Enqueue,
        }
    },

// ... handle other request types

};
```

Reply:
I really like the way this worked out--thanks for suggesting it!

Reply:
glad to hear it! I like the way the new version looks, too.